File: graphite-server/internal/analysis/c4/external.go
package c4

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/johnsonlee/graphite/graphite-server/internal/store"
)

// WeightedClass is ordered evidence. Keeping this as a slice preserves Kotlin
// linked-map tie ordering, which would be lost in a Go map iteration.
type WeightedClass struct {
	ClassName string
	Weight    int
}

func ExternalDependencyKey(graph *store.Store, class string) string {
	if graph != nil {
		if origin, ok := graph.Metadata.ClassOrigins[class]; ok {
			if artifact, ok := ArtifactKey(origin); ok {
				return ArtifactIDPrefix + artifact
			}
		}
	}
	switch {
	case IsJavaRuntimeClass(class):
		return RuntimeIDPrefix + "java"
	case strings.HasPrefix(class, "kotlin."):
		return RuntimeIDPrefix + "kotlin"
	case strings.HasPrefix(class, "scala."):
		return RuntimeIDPrefix + "scala"
	default:
		return NamespaceIDPrefix + NamespaceGroup(class)
	}
}
func ExternalDependencyName(graph *store.Store, key string, classes []string) string {
	switch key {
	case "runtime:java":
		return "Java Runtime"
	case "runtime:kotlin":
		return "Kotlin Runtime"
	case "runtime:scala":
		return "Scala Runtime"
	}
	if strings.HasPrefix(key, ArtifactIDPrefix) {
		return strings.TrimPrefix(key, ArtifactIDPrefix)
	}
	if strings.HasPrefix(key, NamespaceIDPrefix) {
		return strings.TrimPrefix(key, NamespaceIDPrefix)
	}
	if graph != nil {
		for _, class := range classes {
			if origin, ok := graph.Metadata.ClassOrigins[class]; ok {
				return origin
			}
		}
	}
	return key
}
func ExternalDependencySource(key string) string {
	if strings.HasPrefix(key, ArtifactIDPrefix) {
		return "artifact"
	}
	if strings.HasPrefix(key, RuntimeIDPrefix) {
		return "runtime"
	}
	return "namespace"
}
func ClassifyDependencyKind(key string) DependencyKind {
	if strings.HasPrefix(key, RuntimeIDPrefix) {
		return RuntimeDependency
	}
	if strings.HasPrefix(key, ArtifactIDPrefix) {
		return LibraryDependency
	}
	return ExternalSystemDependency
}
func ExternalDependencyConfidence(key string) string {
	if strings.HasPrefix(key, ArtifactIDPrefix) || strings.HasPrefix(key, RuntimeIDPrefix) {
		return "high"
	}
	return "medium"
}
func ExternalDependencyResponsibility(kind DependencyKind) string {
	switch kind {
	case RuntimeDependency:
		return "Provides language and platform runtime services used by the application"
	case LibraryDependency:
		return "Provides reusable library capabilities linked from the application runtime"
	default:
		return "Represents an inferred external software system boundary grouped from referenced classes"
	}
}
func ExternalDependencyDescription(kind DependencyKind) string {
	switch kind {
	case RuntimeDependency:
		return "Language and platform runtime supporting the subject system"
	case LibraryDependency:
		return "Reusable third-party library capabilities referenced from the subject system"
	case ExternalSystemDependency:
		return "External software system candidate inferred from referenced-but-absent classes"
	default:
		return "External collaborator inferred from code graph evidence"
	}
}
func ArtifactKey(origin string) (string, bool) {
	candidate := strings.TrimRight(strings.TrimSpace(origin), "/")
	if i := strings.LastIndexByte(candidate, '/'); i >= 0 {
		candidate = candidate[i+1:]
	}
	candidate = strings.TrimSuffix(candidate, ".jar")
	return candidate, strings.TrimSpace(candidate) != ""
}
func ArtifactNameFromDependencyID(id string) (string, bool) {
	key := strings.TrimPrefix(id, DependencyIDPrefix)
	if !strings.HasPrefix(key, ArtifactIDPrefix) {
		return "", false
	}
	name := strings.TrimPrefix(key, ArtifactIDPrefix)
	return name, strings.TrimSpace(name) != ""
}

var artifactVersion = regexp.MustCompile(`-\d+(?:[.-][0-9A-Za-z]+)*$`)

func ArtifactBaseName(name string) string { return artifactVersion.ReplaceAllString(name, "") }
func NamespaceGroup(name string) string {
	segments := nonblankSegments(name)
	if len(segments) == 0 {
		return name
	}
	cutoff := -1
	for i, segment := range segments {
		first, _ := utf8.DecodeRuneInString(segment)
		if first <= 0xffff && unicode.IsUpper(first) {
			cutoff = i
			break
		}
	}
	packages := segments
	if cutoff > 0 {
		packages = segments[:cutoff]
	}
	root := NamespaceRootSegmentCount(packages)
	selected := min(3, len(packages))
	if root == 1 {
		selected = root
	}
	return strings.Join(packages[:selected], ".")
}
func namespaceFamily(key string) string {
	parts := nonblankSegments(strings.TrimPrefix(key, NamespaceIDPrefix))
	return strings.Join(parts[:min(3, len(parts))], ".")
}
func SummarizeExternalDependencies(graph *store.Store, weights []WeightedClass, limit int) ([]ExternalDependency, error) {
	if limit < 0 {
		return nil, fmt.Errorf("requested element count %d is less than zero", limit)
	}
	grouped := map[string][]WeightedClass{}
	order := []string{}
	for _, entry := range weights {
		key := ExternalDependencyKey(graph, entry.ClassName)
		if _, exists := grouped[key]; !exists {
			order = append(order, key)
		}
		grouped[key] = append(grouped[key], entry)
	}
	families := map[string]int{}
	for _, key := range order {
		if strings.HasPrefix(key, NamespaceIDPrefix) {
			families[namespaceFamily(key)]++
		}
	}
	merged := map[string][]WeightedClass{}
	mergedOrder := []string{}
	for _, key := range order {
		mergedKey := key
		if strings.HasPrefix(key, NamespaceIDPrefix) {
			family := namespaceFamily(key)
			if families[family] > 1 {
				mergedKey = NamespaceIDPrefix + family
			}
		}
		if _, exists := merged[mergedKey]; !exists {
			mergedOrder = append(mergedOrder, mergedKey)
		}
		merged[mergedKey] = append(merged[mergedKey], grouped[key]...)
	}
	out := make([]ExternalDependency, 0, len(mergedOrder))
	for _, key := range mergedOrder {
		weight := 0
		classes := []string{}
		for _, entry := range merged[key] {
			weight += entry.Weight
			classes = append(classes, entry.ClassName)
		}
		kind := ClassifyDependencyKind(key)
		out = append(out, ExternalDependency{ID: DependencyIDPrefix + key, Name: ExternalDependencyName(graph, key, classes), Weight: weight, Source: ExternalDependencySource(key), Kind: kind, Confidence: ExternalDependencyConfidence(key), Responsibility: ExternalDependencyResponsibility(kind), Artifacts: []string{}})
	}
	sort.SliceStable(out, func(i, j int) bool { return out[i].Weight > out[j].Weight })
	return out[:min(limit, len(out))], nil
}

File: rust/graphite-explore-rs/src/c4/external.rs
//! External dependency classification: artifact, runtime or namespace grouping.

use super::boundary::*;
use super::constants::*;
use graphite_storage::Graph;
use indexmap::IndexMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DependencyKind {
    Runtime,
    Library,
    ExternalSystem,
}

impl DependencyKind {
    pub fn wire(self) -> &'static str {
        match self {
            DependencyKind::Runtime => "runtime",
            DependencyKind::Library => "library",
            DependencyKind::ExternalSystem => "external-system",
        }
    }
    pub fn architecture_type(self) -> &'static str {
        match self {
            DependencyKind::Runtime => "runtime-platform",
            DependencyKind::Library => "external-library",
            DependencyKind::ExternalSystem => "external-system",
        }
    }
    pub fn description(self) -> &'static str {
        match self {
            DependencyKind::Library => {
                "Reusable third-party library capabilities referenced from the subject system"
            }
            DependencyKind::ExternalSystem => {
                "External software system candidate inferred from referenced-but-absent classes"
            }
            DependencyKind::Runtime => {
                "Language and platform runtime supporting the subject system"
            }
        }
    }
    pub fn responsibility(self) -> &'static str {
        match self {
            DependencyKind::Runtime => {
                "Provides language and platform runtime services used by the application"
            }
            DependencyKind::Library => {
                "Provides reusable library capabilities linked from the application runtime"
            }
            DependencyKind::ExternalSystem => {
                "Represents an inferred external software system boundary grouped from referenced classes"
            }
        }
    }
}

#[derive(Debug, Clone)]
pub struct ExternalDependency {
    pub id: String,
    pub name: String,
    pub weight: i64,
    pub source: String,
    pub kind: DependencyKind,
    pub confidence: String,
    pub responsibility: String,
    /// Member artifacts when several collapse into one family entry.
    #[allow(dead_code)]
    pub artifacts: Vec<String>,
}

/// `lib/postgresql-42.7.3.jar` -> `postgresql-42.7.3`.
pub fn artifact_key(origin: &str) -> Option<String> {
    let candidate = origin
        .trim()
        .trim_end_matches('/')
        .rsplit('/')
        .next()
        .unwrap_or("")
        .strip_suffix(".jar")
        .unwrap_or_else(|| {
            origin
                .trim()
                .trim_end_matches('/')
                .rsplit('/')
                .next()
                .unwrap_or("")
        });
    if candidate.trim().is_empty() {
        None
    } else {
        Some(candidate.to_string())
    }
}

/// Group a class name down to its namespace, stopping before the first type segment.
pub fn namespace_group(name: &str) -> String {
    let segments: Vec<&str> = name.split('.').filter(|s| !s.is_empty()).collect();
    if segments.is_empty() {
        return name.to_string();
    }
    let cutoff = segments
        .iter()
        .position(|s| s.chars().next().map(|c| c.is_uppercase()).unwrap_or(false));
    let package_segments: &[&str] = match cutoff {
        Some(i) if i > 0 => &segments[..i],
        _ => &segments,
    };
    let root = namespace_root_segment_count(package_segments);
    let take = if root == 1 {
        1
    } else {
        DEFAULT_SEGMENTS.min(package_segments.len())
    };
    package_segments[..take].join(".")
}

/// Stable identity for an external class: artifact, then runtime, then namespace.
pub fn key(g: &Graph, class_name: &str) -> String {
    if let Some(origin) = g.class_origin(class_name) {
        if let Some(k) = artifact_key(origin) {
            return format!("{ARTIFACT_ID_PREFIX}{k}");
        }
    }
    if is_java_runtime_class(class_name) {
        return format!("{RUNTIME_ID_PREFIX}java");
    }
    if class_name.starts_with("kotlin.") {
        return format!("{RUNTIME_ID_PREFIX}kotlin");
    }
    if class_name.starts_with("scala.") {
        return format!("{RUNTIME_ID_PREFIX}scala");
    }
    format!("{NAMESPACE_ID_PREFIX}{}", namespace_group(class_name))
}

fn namespace_family(k: &str) -> String {
    k.strip_prefix(NAMESPACE_ID_PREFIX)
        .unwrap_or(k)
        .split('.')
        .filter(|s| !s.is_empty())
        .take(FAMILY_SEGMENTS)
        .collect::<Vec<_>>()
        .join(".")
}

fn dependency_kind(k: &str) -> DependencyKind {
    if k.starts_with(RUNTIME_ID_PREFIX) {
        DependencyKind::Runtime
    } else if k.starts_with(ARTIFACT_ID_PREFIX) {
        DependencyKind::Library
    } else {
        DependencyKind::ExternalSystem
    }
}

fn source_of(k: &str) -> &'static str {
    if k.starts_with(ARTIFACT_ID_PREFIX) {
        "artifact"
    } else if k.starts_with(RUNTIME_ID_PREFIX) {
        "runtime"
    } else {
        "namespace"
    }
}

fn confidence_of(k: &str) -> &'static str {
    if k.starts_with(ARTIFACT_ID_PREFIX) || k.starts_with(RUNTIME_ID_PREFIX) {
        "high"
    } else {
        "medium"
    }
}

fn name_of(g: &Graph, k: &str, class_names: &[String]) -> String {
    match k {
        "runtime:java" => return "Java Runtime".into(),
        "runtime:kotlin" => return "Kotlin Runtime".into(),
        "runtime:scala" => return "Scala Runtime".into(),
        _ => {}
    }
    if let Some(rest) = k.strip_prefix(ARTIFACT_ID_PREFIX) {
        return rest.to_string();
    }
    if let Some(rest) = k.strip_prefix(NAMESPACE_ID_PREFIX) {
        return rest.to_string();
    }
    class_names
        .iter()
        .find_map(|c| g.class_origin(c))
        .map(|s| s.to_string())
        .unwrap_or_else(|| k.to_string())
}

/// Collapse per-class weights into dependencies, merging namespace families that
/// contribute more than one distinct namespace.
pub fn summarize(g: &Graph, external_weights: &IndexMap<String, i64>) -> Vec<ExternalDependency> {
    let mut grouped: IndexMap<String, Vec<(String, i64)>> = IndexMap::new();
    for (class_name, weight) in external_weights {
        grouped
            .entry(key(g, class_name))
            .or_default()
            .push((class_name.clone(), *weight));
    }
    let mut family_counts: IndexMap<String, usize> = IndexMap::new();
    for k in grouped.keys() {
        if k.starts_with(NAMESPACE_ID_PREFIX) {
            *family_counts.entry(namespace_family(k)).or_insert(0) += 1;
        }
    }
    let mut merged: IndexMap<String, Vec<(String, i64)>> = IndexMap::new();
    for (k, entries) in grouped {
        let target = if k.starts_with(NAMESPACE_ID_PREFIX) {
            let family = namespace_family(&k);
            if family_counts.get(&family).copied().unwrap_or(0) > 1 {
                format!("{NAMESPACE_ID_PREFIX}{family}")
            } else {
                k.clone()
            }
        } else {
            k.clone()
        };
        merged.entry(target).or_default().extend(entries);
    }
    let mut out: Vec<ExternalDependency> = merged
        .into_iter()
        .map(|(k, entries)| {
            let class_names: Vec<String> = entries.iter().map(|(c, _)| c.clone()).collect();
            let kind = dependency_kind(&k);
            ExternalDependency {
                id: format!("{DEPENDENCY_ID_PREFIX}{k}"),
                name: name_of(g, &k, &class_names),
                weight: entries.iter().map(|(_, w)| *w).sum(),
                source: source_of(&k).to_string(),
                kind,
                confidence: confidence_of(&k).to_string(),
                responsibility: kind.responsibility().to_string(),
                artifacts: Vec::new(),
            }
        })
        .collect();
    // Stable sort keeps insertion order for equal weights.
    out.sort_by(|a, b| b.weight.cmp(&a.weight));
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn artifact_key_strips_path_and_extension() {
        assert_eq!(
            artifact_key("lib/postgresql-42.7.3.jar").unwrap(),
            "postgresql-42.7.3"
        );
        assert_eq!(artifact_key("BOOT-INF/classes/").unwrap(), "classes");
        assert!(artifact_key("").is_none());
    }

    #[test]
    fn namespace_group_stops_before_the_type_segment() {
        assert_eq!(
            namespace_group("com.partner.payment.PaymentGateway"),
            "com.partner.payment"
        );
        assert_eq!(
            namespace_group("okhttp3.internal.connection.RealConnection"),
            "okhttp3"
        );
    }

    #[test]
    fn dependency_kinds_follow_the_id_prefix() {
        assert_eq!(dependency_kind("artifact:x"), DependencyKind::Library);
        assert_eq!(dependency_kind("runtime:java"), DependencyKind::Runtime);
        assert_eq!(
            dependency_kind("namespace:com.x"),
            DependencyKind::ExternalSystem
        );
    }
}
